trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java

```diff
@@ -1301,10 +1298,123 @@ private Result handleAtomicExchange(final IDispatcher main, final IASTFunctionCa
+				mExprResultTransformer.transformDispatchSwitchRexBoolToInt(main, loc, arguments[5]);
+		final var resultBuilder = new ExpressionResultBuilder().addAllExceptLrValue(pointer, expectedResult,
+				desiredResult, weakResult, successMemoryOrder, failureMemoryOrder);
+
+		// Introduce an auxvar indicating whether the function is successful, i.e., the exchange was performed.
+		final var boolType = new CPrimitive(CPrimitives.BOOL);
+		final var success = mAuxVarInfoBuilder.constructAuxVarInfo(loc, boolType, AUXVAR.RETURNED);
+		final var boolZero = mExpressionTranslation.constructZero(loc, boolType);
+		final var successBoolean = mExpressionTranslation.constructBinaryEqualityExpression(loc,
+				IASTBinaryExpression.op_notequals, success.getExp(), boolType, boolZero, boolType);
+		final var noSuccessBoolean = mExpressionTranslation.constructBinaryEqualityExpression(loc,
+				IASTBinaryExpression.op_equals, success.getExp(), boolType, boolZero, boolType);
+
+		final var pointerRead = mExprResultTransformer.readPointerValue(loc, pointer.getLrValue());
+		final var expectedRead = mExprResultTransformer.readPointerValue(loc, expectedResult.getLrValue());
+		final var pointerWrite = mExprResultTransformer.makePointerAssignment(loc, pointer.getLrValue(),
+				desiredRead.getLrValue().getValue());
+
+		final var atomicBody = new ExpressionResultBuilder().addAuxVarWithDeclaration(success)
+				.addStatement(StatementFactory.constructIfStatement(loc, weakResult.getLrValue().getValue(),
+						new Statement[] { new HavocStatement(loc, new VariableLHS[] { success.getLhs() }) },
+						new Statement[] { StatementFactory.constructSingleAssignmentStatement(loc, success.getLhs(),
+								mExpressionTranslation.constructLiteralForIntegerType(loc, boolType,
+										BigInteger.ONE)) }))
+				.addAllExceptLrValue(pointerRead, expectedRead).addAllExceptLrValueAndStatements(desiredRead)
+				.addAllExceptLrValueAndStatements(pointerWrite)
+				.addStatement(StatementFactory.constructIfStatement(loc,
+						new BinaryExpression(loc, Operator.LOGICAND, successBoolean,
+								new BinaryExpression(loc, Operator.COMPEQ, pointerRead.getLrValue().getValue(),
+										expectedRead.getLrValue().getValue())),
+						DataStructureUtils.concat(desiredRead.getStatements(), pointerWrite.getStatements())
+								.toArray(Statement[]::new),
+						new Statement[] { StatementFactory.constructSingleAssignmentStatement(loc, success.getLhs(),
+								mExpressionTranslation.constructLiteralForIntegerType(loc, boolType,
+										BigInteger.ZERO)) }))
+				.build();
+		final var atomic = applyMemoryOrders(loc, atomicBody, successMemoryOrder.getLrValue().getValue(),
+				failureMemoryOrder.getLrValue().getValue());
+
+		final var expectedWrite = mExprResultTransformer.makePointerAssignment(loc, expectedResult.getLrValue(),
+				pointerRead.getLrValue().getValue());
+		return resultBuilder.addAllExceptLrValue(atomic).addAllExceptLrValueAndStatements(expectedWrite)

```

Comment:
Is it correct that the write is indeed not atomic? I don't fully understand this from the specification.

Reply:
Hmm, I'm not sure. I was basing this on the description at https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html#index-_005f_005fatomic_005fcompare_005fexchange_005fn which says:

> If equal, the operation is a _read-modify-write_ operation that writes `desired` into `*ptr`. If they are not equal, the operation is a _read_ and the current contents of `*ptr` are written into `*expected.`

I understood that to say that the write is at least not atomic with the rest, otherwise the operation would be a _read-modify-write_. Though that still leaves open whether the write is atomic by itself.

The description of `atomic_compare_exchange` in Section 7.17.7.4 of the C standard sounds somewhat different: 
![image](https://github.com/user-attachments/assets/b4f74ca7-d6d8-4fc3-a20a-249fa502608d)

This seems more like the failure case is atomic with the rest. Of course, we're implementing the GNU variant, which "approximates" the standard... :shrug:

Reply:
The C standard is also a bit more precise what "fail spuriously" looks like; it's different from the behaviour for different values. I will have to update the code for that as well.
![image](https://github.com/user-attachments/assets/b44845a8-1a35-45f0-bdc1-6060ae2bde20)

Reply:
I've reimplemented the method based on the C standard description.